src/nemo_evaluator/observability/failure_classification.py

```
@@ -0,0 +1,73 @@
+from typing import Any
+
+MODEL_FAILURE_CATEGORIES = frozenset({"budget_exceeded", "model_timeout", "rate_limit", "server_error", "model_error"})
+
+_STATUS_RE = re.compile(
+    r"(?:status[_ -]?code\s*[=:]\s*|error code:\s*|http\s+|status\s+)(?P<code>[1-5][0-9]{2})\b",
+    re.IGNORECASE,
+)
+_LEADING_STATUS_RE = re.compile(r"^\D*(?P<code>[1-5][0-9]{2})\b")
+
+
+def _status_int(value: Any) -> int | None:
```

Ruff (0.15.20)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `value`

(ANN401)